lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -25,16 +25,228 @@
+      // Record all the hierarchy names.
+      SmallVector<std::string> hierNames;
+      J.attributeArray("hierarchy", [&] {
+        for (auto p : paths) {
+          const InstanceOp &x = p.front();
+          std::string hierName =
+              x->getParentOfType<FModuleOp>().getName().str();
+          for (InstanceOp inst : p) {
+            hierName = hierName + "." + inst.name().str();
+          }
+          hierNames.push_back(hierName);
+          J.value(hierName);
+        }
+      });
+      // If verification annotation added to the memory op then print the data.
+      if (verifData)
+        J.attributeObject("verification_only_data", [&] {
+          for (auto name : hierNames) {
+            J.attributeObject(name, [&] {
+              for (auto data : verifData) {
+                // Id for the memory verification property.
+                std::string id = data.first.strref().str();
+                // Value for the property.
+                auto verifValAttr = data.second;
+                // Now print the value attribute based on its type.
+                printAttr(J, verifValAttr, id);
+              }
+            });
+          }
+        });
+    });
+  };
+  std::string testBenchJsonBuffer;
+  llvm::raw_string_ostream testBenchOs(testBenchJsonBuffer);
+  llvm::json::OStream testBenchJson(testBenchOs);
+  std::string dutJsonBuffer;
+  llvm::raw_string_ostream dutOs(dutJsonBuffer);
+  llvm::json::OStream dutJson(dutOs);
+  SmallVector<MemOp> dutMems;
+  SmallVector<MemOp> tbMems;
+
+  for (auto mod : circuitOp.getOps<FModuleOp>()) {
+    bool isDut = dutModuleSet.contains(mod);
+    for (auto memOp : mod.getBody().getOps<MemOp>())
+      if (isDut)
+        dutMems.push_back(memOp);
+      else
+        tbMems.push_back(memOp);
+  }
+  std::string seqMemConfStr, tbConfStr;
+  dutJson.array([&] {
+    for (auto memOp : dutMems)
+      createMemMetadata(memOp, dutJson, seqMemConfStr);
+  });
+  testBenchJson.array([&] {
+    for (auto memOp : tbMems)
+      createMemMetadata(memOp, testBenchJson, tbConfStr);
+  });
+
+  auto *context = &getContext();
+  auto builder = OpBuilder::atBlockEnd(circuitOp.getBody());
+  AnnotationSet annos(circuitOp);
+  auto diranno = annos.getAnnotation(metadataDirectoryAttrName);
+  std::string metadataDir = "metadata";
+  if (diranno) {
+    if (auto dir = diranno.get("dirname")) {
+      if (dir.isa<StringAttr>())
+        metadataDir = dir.cast<StringAttr>().getValue().str();
+    }
+  }
+  // Use unknown loc to avoid printing the location in the metadata files.
+  auto tbVerbatimOp = builder.create<sv::VerbatimOp>(builder.getUnknownLoc(),
+                                                     testBenchJsonBuffer);
+  auto dutVerbatimOp =
+      builder.create<sv::VerbatimOp>(builder.getUnknownLoc(), dutJsonBuffer);
+  auto confVerbatimOp =
+      builder.create<sv::VerbatimOp>(builder.getUnknownLoc(), seqMemConfStr);
+  if (!metadataDir.empty()) {
```

Comment:
Could an empty string be used to output to the current directory?

Reply:
Right, the dir could be specified as empty. I think we can remove the if check.